src/garage/torch/algos/maml.py

```diff
@@ -0,0 +1,475 @@
+"""Model-Agnostic Meta-Learning (MAML) algorithm implementation for RL."""
+import collections
+import copy
+
+from dowel import tabular
+import numpy as np
+import torch
+
+from garage.misc import tensor_utils
+from garage.sampler import OnPolicyVectorizedSampler
+from garage.tf.samplers import BatchSampler
+from garage.torch.algos import _Default, make_optimizer, VPG
+from garage.torch.optimizers import DiffSGD
+from garage.torch.utils import update_module_params
+
+
+class MAML:
+    """Model-Agnostic Meta-Learning (MAML).
+
+    Args:
+        env (garage.envs.GarageEnv): A gym environment.
+        policy (garage.torch.policies.Policy): Policy.
+        baseline (garage.np.baselines.Baseline): The baseline.
+        meta_optimizer (Union[torch.optim.Optimizer, tuple]):
+            Type of optimizer.
+            This can be an optimizer type such as `torch.optim.Adam` or a tuple
+            of type and dictionary, where dictionary contains arguments to
+            initialize the optimizer e.g. `(torch.optim.Adam, {'lr' = 1e-3})`.
+        meta_batch_size (int): Number of tasks sampled per batch.
+        inner_lr (double): Adaptation learning rate.
+        num_grad_updates (int): Number of adaptation gradient steps.
+        inner_algo (garage.torch.algos.VPG): The inner algorithm used for
+            computing loss.
+
+    """
+
+    def __init__(self,
+                 env,
+                 policy,
+                 baseline,
+                 meta_optimizer=torch.optim.Adam,
+                 meta_batch_size=40,
+                 inner_lr=0.1,
+                 num_grad_updates=1,
+                 inner_algo=None):
+        if inner_algo is None:
+            inner_algo = VPG(env.spec, policy, baseline)
+
+        if policy.vectorized:
+            self.sampler_cls = OnPolicyVectorizedSampler
+        else:
+            self.sampler_cls = BatchSampler
+
+        self.policy = policy
+        self.max_path_length = inner_algo.max_path_length
+        self._env = env
+        self._baselines = [
+            copy.deepcopy(baseline) for _ in range(meta_batch_size)
```

Comment:
what if `meta_batch_size` != `number of tasks`? does this require that the index of each task in the meta-batch be consistent?

Reply:
I assume that `meta_batch_size` is defined as the number of tasks. Can you explain what is the difference between the two?

Reply:
Thanks for the explanation. 

Prior arts uses one baseline for all `tasks[i][j]`. This is OK because the baseline is not recurrent and fit-before-predict. From previous experiments, I see no performance impact if I apply this to garage. 

Use `meta_batch_size` baselines is to support fit-after-predict. In such a case, using only one baseline implies that we fit on `tasks[i][j]` and predict on `tasks[i][j+1]`, which is wrong.

Reply:
I see. Did you decide to use fit-after-predict or fit-before-predict by default?

Reply:
I wasn't aware supporting both would require such a big architectural change.

If all of the prior art uses fit-before-predict, then I suggest you do the same and eliminate some complexity.

I think the understand the nature of the problem better now, and fit-before-predict is a little bit of a hack to make on-policy meta-RL work. I won't go into details here but feel free to ask me in person.

Reply:
OK. I will use fit-before-predict because its give SOTA performance.

Reply:
LGTM

Reply:
In the case of tf branch, we need n baselines since the sample paths are processed in batch, so we can use the i-th baseline for the i-th task, Unlike ProMP where it loops through the paths and fit/predict one by one.

Reply:
@ahtsan I'm not sure I follow, since the `ith` task isn't necessarily the same task between batches. This would reveal the task ID.

Reply:
I meant the `i-th` environment, i.e. the `i-th` element of `self._vec_env.envs` in the sampler.

Reply:
Do you assume the the `i`-th element of `self._vec_env.envs` corresponds to the same task between batches?

i.e. imagine there are only 5 tasks, with IDs 0-4. Do you assume that that if task 0 appears at `self._vec_env.envs[0]` on iteration 1 that it will also appear at `self._vec_env.envs[0]` on iteration 2?

Reply:
Yes if we only set `self._vec_env.envs[0]` to task 0 at the first place and we did not set it to something else after, since we did not update ``self._vec_env.envs[0]`.
No if we set it to something else for each iteration, and `self._vec_env.envs[0]` will be what it is set to be.

Basically what I did is to align `baselines[i]` with `self._vec_env.envs[i]`, regardless of what the `i-th` task is, so that I can fit and predict for `self._vec_env.envs[i]` using `baselines[i]` in any iteration, without knowing what the `i-th` task is.

Reply:
this sounds fine to me

a slight optimization for RL2 might be to fit a single baseline `global_baseline` with all data at the end of each iteration after policy optimization (`global_baseline.fit(all_paths)`). then, at the beginning of the next iteration, `[baselines[i].parameters = global_baseline.parameters for i in range(meta_batch_size)]`. this allows all of your baselines to share parameters between updates. this might allow your baseline to learn a little faster, especially at the beginning.